Specification of URL to TLD for GUI (#1429)
dotse/zonemaster: docs/public/configuration/tld-url-specification.md

Reviewer: so, i can write `https://nic.tld/do?domain=[DOMAIN]` or `https://nic.tld/zm/[DOMAIN]`  for example,but not `https://[DOMAIN].nic.tld`
We are removing a possibility of redirection for the user, but that seems reasonable to me, it's seems complicated for little gain.

Reviewer: Forget about it. Let's wait and see if this is a user request.

# TLD URL Specification

## Table of contents

* [Introduction](#introduction)
* [Limitations](#limitations)
* [IDN names](#idn-names)
* [Source of URL](#source-of-url)
* [Blocking policy](#blocking-policy)
* [URL string or blocking policy]
  * [Backend configuration](#backend-configuration)
  * [TXT record]
  * [Examples](#examples)
* [URL from IANA RDAP database]

## Introduction

To facilitate redirection of a GUI user to the responsible party for the operation
of a tested domain name, an URL to the relevant TLD is presented with the test
result. This document defines how the URL is set based on public information, and
where in Zonemaster that URL can be changed to another value on a specific

> Zonemaster installation.

Reviewer 2: Here's a stab at making the introductory paragraph more direct. IMHO, this version has a number of merits:
* it refer to the feature by name,
* it talks about the feature as a thing instead of talking around it,
* it focuses on the feature rather than the document, and
* it avoids claiming that the configuration is detailed in this document.

Also, the reason this version is so much shorter is that I moved some details on sources from here to another suggestion a few lines down.

```suggestion
The TLD URL feature provides a way to determine a URL to the TLD closest to a given
domain name. Zonemaster GUI makes use of this feature on its test result pages.
```

How the GUI gets the URL is defined in [Backend RPC API].

How Backend can override public values is defined in [Backend configuration].

> 

Reviewer 2: Good idea to point out where other aspects of the feature are defined. I think we should contrast that with the part played by this document.

```suggestion

This document specifies how the URL or its absence is determined.
The URL is determined based on based on one or more of:

* a given domain name,
* Backend configuration,
* DNS data published by the relevant TLD, and
* RDAP data published by IANA.

```

In this document the only types of [URLs][URL] that are considered are `http`

> and `https`, i.e. URLs where the scheme is `http` or `https`. There are also

Reviewer 2: This spec does in fact check for HTTP(S). The only thing missing is how to deal with failure in that set of checks, but that ought to be clarified anyways.

Author: I do not understand the comment.

Author: I found an update.

restrictions on allowed characters in that URL in section
[URL string or blocking policy]. In the same section the term
`URL string` is used, and it is defined at the start of that section.

> ## Limitations

Reviewer 2: This section both deals with preconditions and determination. IMHO it ought to be split along this line and integrated with the IDN preconditions and the Source of URL determination, respectively.
Knowing when the procedure is even applicable is a greater concern than how it behaves when applicable.

Author: Please elaborate.

The URL used is based on the tested domain name. If the tested domain name
matches one of the following conditions then no URL will be provided, and that can

> not be overridden by Backend configuration.

Reviewer 2: IMHO we should talk about the given domain rather than the tested domain. That is more accurate since the procedure can be invoked without any domains being tested. E.g., by a TLD operator who wants to debug their stuff.

Author: Other times we write "tested domain" or "domain tested". It can be confusing if we here talk about the "given domain". I will look at the text.

* The tested domain name is illegal, i.e. it contains illegal character(s) that
  prevents it from being tested by Zonemaster.
* The tested domain name is the root zone (`.`).
* The tested domain name is a TLD, e.g. `se` or `fr`.

In all other cases, a URL will be provided, if available and permitted by policy configuration.

## IDN names

If the tested domain name has one or more IDN labels submitted in U-label
format, then those must be converted to A-label format before the steps below can

> be run. This includes the TLD label.

Reviewer 2: This section ought to be renamed "Preconditions" and moved up above the "Limitations" section.

Also I don't think it's the right place to talk about requests from clients here. That's the responsibility of the RPCAPI. This document deals with the business logic. It would be more appropriate for this document to state what the business logic expects from its caller. Personally I'd prefer just requiring A-labels here.

Oh, and IMHO we should add the precondition that all labels must be lowercase.

Author: I will look at it.

## Source of URL

> The URL is specifically set on a per TLD basis.

Reviewer 2: We could use this line to declare the inputs and outputs of the procedure. WDYT?
```suggestion
The URL (or its absence) is determined based on public information, configuration and the given TLD.
```

The following priority applies for determining the source of the URL:
* Highest priority is [Backend configuration] if it is configured with
  a specific URL for that TLD. See [Backend configuration] for how to
  configure the TLD specific URL.
* Second priority is a specific DNS record as specified below in section [TXT record].
* The fallback is to fetch the URL for registration services found in the IANA

>   RDAP database as specified below in section [URL from IANA RDAP database].

Reviewer 2: It is possible that the fallback fails to produce a result, so there should be a fallback for the fallback. And I took the liberty to tweak the wording a little bit. WDYT?

```suggestion
1. [Backend configuration] may set a specific URL for the given TLD.
2. A specific DNS record may be published by the given TLD, as specified below in section [TXT record].
3. The URL for registration services found in the IANA
   RDAP database as specified below in section [URL from IANA RDAP database].
4. The fallback is to return absence of a URL.
```

Author: In your proposal only selection of URL is considered, but there is also the "opt out" mechanism.

## Blocking policy

A blocking policy can be defined to prevent an URL from being shown.
If a blocking policy is found then an empty URL is returned, as if no URL was found for the TLD.

The following priority applies for blocking policies:
* Highest priority is in the global policy in the [Backend configuration].
  If the global policy is set to block, then no URLs will be used for
  any domain name of any TLD, i.e. turning the feature off.
* The second priority is also set in the [Backend configuration], but this time
  per TLD. If it is set to block then no URL will be used for that TLD,
  independently of the availability of the URL from any source.
* The last priority can be set by the TLD manager in a TXT record where the URL
  string can be provided. The format and requirements are specified below in
  section [TXT record].
  * If a blocking policy is found then no URL is fetched from the IANA RDAP
    database. However, this blocking policy has no affect on a TLD string
    specified in [Backend configuration].

## URL string or blocking policy

### Terminology

The term `URL string` in used in this section for a string from which a URL can

> be derived using the steps in [TXT record]. Specifically an `URL string` may

Reviewer 2: ```suggestion
be derived using the steps in [TXT record]. Specifically a `URL string` may
```

      * The path string may also contain the literal string `[DOMAIN]` somewhere
        after the first solidus `/`.
      * If the literal string `[DOMAIN]` is found in the path string it will be
        replaced by the tested domain name. If the tested domain name contains
        solidus (slash) `/` then that will be encoded as `%2F`.
      * An empty path string will be replaced by the string `/`.
* If the text string is neither a blocking policy (literal string `[BLOCK]`) or a
  valid URL string, then the DNS TXT record is ignored.

### Examples

Invalid domain names:

* `green.xa.` (must not have trailing full stop `.`)
* `green-.xa` (label must not end with a hyphen-minus `-`)
* `grön.xa` (U-label not permitted, use A-label instead)
* `green_apple.xa` (low line (underscore) `_` is not permitted)

Invalid path strings:

* `/domän` (invalid character `ä`, use ASCII only)
* `/domain=<domain>` (`<domain>` is invalid, use `[DOMAIN]` instead)
* `/domain/search=$` (`$` is invalid)

Valid domain name strings:

* `green.xa`
* `xn--grn-tna.xa` (valid with A-label)
* `green-apple.xa`

Valid path strings:

* `/domain\&search=true`
* `/domain/[DOMAIN]`

> * `/registry\&domain=[DOMAIN]`

Reviewer 2: Backslashes are forbidden.
```suggestion
* `/domain/&search=true`
* `/domain/[DOMAIN]`
* `/registry/&domain=[DOMAIN]`
```

Author: Backslashes were typing mistakes, not intentional.

URL from URL string in a TXT record:

```
Tested domain: green.xa
TLD: xa
URL string: https://domain.nic.xa/search/[DOMAIN]
URL: https://domain.nic.xa/search/green.xa
```

```
Tested domain: green.xb
TLD: xb
URL string: https://domain.nic.xa/search/
URL: https://domain.nic.xa/search/
```

```
Tested domain: green.xc
TLD: xc
URL string: https://domain.nic.xa
URL: https://domain.nic.xa/
```

## URL from IANA RDAP database

If the following conditions are true, then a lookup of the URL for the TLD will be done from the

> IANA RDAP database:

Reviewer 2: IMHO the inputs and outputs of these sub-procedures, as well as their data types, should be declared up front.

Author: Will check.

* [Backend configuration] has no global blocking policy.
* [Backend configuration] has no TLD blocking policy.
* [Backend configuration] does not configure a TLD specific URL.
* There is no valid TLD specific TXT record with blocking policy (see
  [URL string or blocking policy]).
* There is no valid TLD specific TXT record that configures a URL (see
  [URL string or blocking policy]).

The base URL for the IANA RDAP database is `https://rdap.iana.org/domain/` to
which the TLD in question is appended, and from which the relevant string can be
extracted using the command below (both `curl` and `jq` must be installed) where

> `na` is used as an example TLD:

Reviewer 2: There's a lot going on in this sentence. Maybe split it up into more focused parts?

Author: Will check.

```sh
curl -s https://rdap.iana.org/domain/na | jq -r '.links[] | select(.rel=="related") | .href'
```

> * The URL fetched must consist of the following parts in that order:

Reviewer 2: I'm reading this as a specification of what URLs to accept, but the algorithm doesn't specify what happens when the URL is rejected. One could assume that the URL should just be discarded, but maybe that should be spelled out.

Also, there seems to be assumptions that the above request always gives a response, and that a single URL is left after collection and filtering. We should probably clarify the behavior when these assumptions do not hold.

Author: I will check.

Reviewer 2: What about the case when jq returns multiple URLs?

  * a literal `https://` or `http://`,
  * a domain name,
  * a path string.
 * The domain name string may contain characters `a-z0-9.-` where
   * full stop (dot) `.` must not be the first or last character,
   * there must not be a sequence of two or more full stops `.`,
   * hyphen-minus `-` must not start or end a label,
   * IDN labels must be represented in the A-label form.
 * The path string may be empty or must start with a solidus (slash) `/` and
      may contain characters `a-zA-Z0-9/=?%_.&-`.
      * An empty path string will be replaced by the string `/`.

This process will extract the same URL as the one for 
"URL for registration services" found in the [IANA Root Zone Database] after

> selecting the relevant TLD.

Reviewer 2: This reads to me like a purpose description for the procedure. IMHO this should be promoted to be the very first paragraph of the section.

Also, is there some API reference or other source we could refer to back up this claim?

And perhaps we should say "should" instead of "will". After all it relies on things that are out of our control.

Author: I will check.

matches one of the following conditions then no URL will be provided, and that
can not be overridden by Backend configuration.

* The given domain name is not normalized.
* The given domain name is the root zone (`.`).
* The given domain name is a TLD, e.g. `se` or `fr`.

## Limitations

The only types of [URLs][URL] that the mechanisms specified in this document can
result in are `http` and `https`, i.e. URLs where the scheme is `http` or
`https`. There are also restrictions on allowed characters in that URL in section
[URL string or blocking policy]. In the same section the term `URL string` is
used, and it is defined at the start of that section.

## Determination of URL

The URL is determined based on public information, configuration and the TLD of
the given domain name. The determination order is as follows:

1. [Backend configuration] may set a specific URL for the given TLD.
2. A specific DNS record may be published by the given TLD, as specified below in
   section [TXT record], with the URL to be used.
3. The URL for registration services found in the IANA RDAP database as specified
   below in section [URL from IANA RDAP database].

> 4. The fallback is to return the absense of a URL.

Reviewer 3: ```suggestion
4. The fallback is to return the absence of a URL.
```

Both in the [Backend configuration] and the [TXT record] there may be a blocking
policy to prevent any URL from being returned. See details in the sections below.

## Blocking policy

A blocking policy can be defined to prevent an URL from being shown.
If a blocking policy is found then an empty URL is returned, as if no URL was found for the TLD.

The following priority applies for blocking policies:
* Highest priority is in the global policy in the [Backend configuration].
  If the global policy is set to block, then no URLs will be used for
  any domain name of any TLD, i.e. turning the feature off.
* The second priority is also set in the [Backend configuration], but this time
  per TLD. If it is set to block then no URL will be used for that TLD,
  independently of the availability of the URL from any source.
* The last priority can be set by the TLD manager in a TXT record where the URL
  string can be provided. The format and requirements are specified below in
  section [TXT record].
  * If a blocking policy is found then no URL is fetched from the IANA RDAP

>     database. However, this blocking policy has no affect on a TLD string

Reviewer 3: ```suggestion
    database. However, this blocking policy has no effect on a TLD string
```

# TLD URL Specification

## Table of contents

* [Introduction](#introduction)
* [Given Domain Name]
* [Limitations](#limitations)

> * [Preconditions](#preconditions)

Reviewer 3: ```suggestion
* [Preconditions](#preconditions)
* [Limitations](#limitations)
```

* [Determination of URL](#determination-of-url)
* [Blocking policy](#blocking-policy)
* [URL string or blocking policy]
  * [Terminology](#terminology)
  * [Backend configuration](#backend-configuration)
  * [TXT record]
  * [Examples](#examples)
* [URL from IANA RDAP database]

## Introduction

The TLD URL feature provides a way to determine a URL to the TLD closest to a given
domain name. Zonemaster GUI makes use of this feature on its test result pages.

How the GUI gets the URL is defined in [Backend RPC API].

How Backend can override public values is defined in [Backend configuration].

This document specifies how the URL or its absence is determined.

> The URL is determined based on based on one or more of:

Reviewer 3: ```suggestion
The URL is determined based on one or more of:
```

* The given domain name,
* Backend configuration,
* DNS data published by the relevant TLD, and
* RDAP data published by IANA.

## Given Domain Name

In this document it is referred to the `given domain name`. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be executed on any domain name whether or not a Zonemaster test is

> executed on the domain name.

Reviewer 3: ```suggestion
In this document the term `given domain name` is used. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be used on any domain name whether or not a Zonemaster test is
executed on the domain name.
```

The URL used is based on the [given domain name][#given-domain-name]. If the
given domain name matches one of the following conditions then no URL will be
provided, and that can not be overridden by Backend configuration.

* The given domain name is illegal, i.e. it contains illegal character(s) that
  [prevents it from being tested by Zonemaster][Requirements]
* The given domain name is the root zone (`.`).
* The given domain name is a TLD, e.g. `se` or `fr`.

In all other cases, a URL will be provided, if available and permitted by policy
configuration.

## Determination of URL

The URL is determined based on public information, configuration and the TLD of
the given domain name. The determination order is as follows:

1. [Backend configuration] may set a specific URL for the given TLD.
2. A specific DNS record may be published by the given TLD, as specified below in
   section [TXT record], with the URL to be used.
3. The URL for registration services found in the IANA RDAP database as specified
   below in section [URL from IANA RDAP database].
5. The fallback is to return the absense of a URL.

Both in the [Backend configuration] and the [TXT record] there may be blocking

> policy to prevent any URL to returned. See details in the sections below.

Reviewer 3: ```suggestion
Both in the [Backend configuration] and the [TXT record] there may be a blocking
policy to prevent any URL from being returned. See details in the sections below.
```

## Blocking policy

A blocking policy can be defined to prevent an URL from being shown.
If a blocking policy is found then an empty URL is returned, as if no URL was found for the TLD.

The following priority applies for blocking policies:
* Highest priority is in the global policy in the [Backend configuration].
  If the global policy is set to block, then no URLs will be used for
  any domain name of any TLD, i.e. turning the feature off.
* The second priority is also set in the [Backend configuration], but this time
  per TLD. If it is set to block then no URL will be used for that TLD,
  independently of the availability of the URL from any source.
* The last priority can be set by the TLD manager in a TXT record where the URL
  string can be provided. The format and requirements are specified below in
  section [TXT record].
  * If a blocking policy is found then no URL is fetched from the IANA RDAP
    database. However, this blocking policy has no affect on a TLD string
    specified in [Backend configuration].

## URL string or blocking policy

### Terminology

> The term `URL string` in used in this section for a string from which a URL can

Reviewer 3: ```suggestion
The term `URL string` used in this section stands for a string from which a URL can
```

> be derived using the steps in [TXT record]. Specifically a `URL string` may

Reviewer 3: ```suggestion
be derived using the steps in [TXT record]. Specifically, a `URL string` may
```

> contain the literal string `[DOMAIN]` which is replaced by the given domain name

Reviewer 3: ```suggestion
contain the literal string `[DOMAIN]` which is to be replaced by the given domain name
```

> when the URL is derived. A `URL` derived from a `URL string` is also a valid

Reviewer 3: ```suggestion
when the URL is derived. A URL derived from a `URL string` is also a valid
```

### Examples

Invalid domain names:

* `green.xa.` (must not have trailing full stop `.`)
* `green-.xa` (label must not end with a hyphen-minus `-`)
* `grön.xa` (U-label not permitted, use A-label instead)
* `green_apple.xa` (low line (underscore) `_` is not permitted)

Invalid path strings:

* `/domän` (invalid character `ä`, use ASCII only)
* `/domain=<domain>` (`<domain>` is invalid, use `[DOMAIN]` instead)
* `/domain/search=$` (`$` is invalid)

Valid domain name strings:

* `green.xa`
* `xn--grn-tna.xa` (valid with A-label)
* `green-apple.xa`

Valid path strings:

* `/domain/&search=true`
* `/domain/[DOMAIN]`
* `/registry/&domain=[DOMAIN]`

URL from URL string in a TXT record:

```
Given domain: green.xa
TLD: xa
URL string: https://domain.nic.xa/search/[DOMAIN]
URL: https://domain.nic.xa/search/green.xa
```

```
Given domain: green.xb
TLD: xb
URL string: https://domain.nic.xa/search/
URL: https://domain.nic.xa/search/
```

```
Given domain: green.xc
TLD: xc
URL string: https://domain.nic.xa
URL: https://domain.nic.xa/
```

## URL from IANA RDAP database

If the publication of the URL was not blocked in the steps above and no URL

> was not determined from the steps above, then a lookup of the URL for the TLD

Reviewer 3: ```suggestion
was determined from the steps above, then a lookup of the URL for the TLD
```

will be done from the IANA RDAP database.

The base URL for the IANA RDAP database is `https://rdap.iana.org/domain/`.

> Append the the given TLD to the base URL. From a lookup of the resulting RDAP

Reviewer 3: ```suggestion
The base URL for the IANA RDAP database is `https://rdap.iana.org/domain/`,
to which the appropriate TLD is appended. From a lookup of the resulting RDAP
```

URL, the URL for the registration service for the given TLD can be found, if
defined.

E.g use the following the command (both `curl` and `jq` must be installed) where

> `na` is used as an example TLD:

Reviewer 3: ```suggestion
For example, use the following command where `na` is used as an example TLD:

> [!NOTE]
> Both `curl` and `jq` must be installed
```

```sh
curl -s https://rdap.iana.org/domain/na | jq -r '.links[] | select(.rel=="related") | .href'
```

> * The URL fetched must consist of the following parts in that order:

Reviewer 3: ```suggestion
* The fetched URL must consist of the following parts, in that order:
```

  * a literal `https://` or `http://`,
  * a domain name,
  * a path string.
 * The domain name string may contain characters `a-z0-9.-` where
   * full stop (dot) `.` must not be the first or last character,
   * there must not be a sequence of two or more full stops `.`,
   * hyphen-minus `-` must not start or end a label,
   * IDN labels must be represented in the A-label form.
 * The path string may be empty or must start with a solidus (slash) `/` and
      may contain characters `a-zA-Z0-9/=?%_.&-`.
      * An empty path string will be replaced by the string `/`.

This process will extract the same URL as the one for "URL for registration
services" found in the [IANA Root Zone Database] after selecting the relevant
TLD.

> If the no URL was found or no URL matching the requirements, then no URL is

Reviewer 3: ```suggestion
If no URL was found or no URL matched the requirements, then no URL is
```

In this document it is referred to the `given domain name`. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be executed on any domain name whether or not a Zonemaster test is
In this document the term `given domain name` is used. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be used on any domain name whether or not a Zonemaster test is
executed on the domain name.

## Preconditions

If the [given domain name][#given-domain-name] has one or more IDN labels
submitted in U-label format, then those must be converted to
[A-label][RFC 5890#2.3.2.1] format before the steps below can be run. This also
applies to the TLD label.

The given domain name must be submitted in lower case, i.e. `A-Z` must be
downcased to `a-z` before submission.

## Limitations

The only types of [URLs][URL] that the mechanisms specified in this document can
result in are `http` and `https`, i.e. URLs where the scheme is `http` or
`https`. There are also restrictions on allowed characters in that URL in section
[URL string or blocking policy]. In the same section the term `URL string` is

> used, and it is defined at the start of that section.

Reviewer 2: This document specifies how the URL or its absence is determined, and there's a separate document for the configuration format. In this light, restrictions on what is accepted by the configuration format are somewhat out of place here. Maybe remove these remarks here? And possibly add them to the configuration document if they aren't already clear enough there.

Author: The configuration is for backend, but this text is for those who want to create a TXT record in the TLD. They should only have to read this document, as I see it.

The URL used is based on the [given domain name][#given-domain-name]. If the
given domain name matches one of the following conditions then no URL will be
provided, and that can not be overridden by Backend configuration.

* The given domain name is illegal, i.e. it contains illegal character(s) that
  [prevents it from being tested by Zonemaster][Requirements]
* The given domain name is the root zone (`.`).

> * The given domain name is a TLD, e.g. `se` or `fr`.

Reviewer 2: The preconditions section deals with input validation of the given domain name. Maybe move these rules up there as well so that everything can be found in the same place?

Author: OK.

# TLD URL Specification

## Table of contents

* [Introduction](#introduction)
* [Given Domain Name]
* [Preconditions]
* [Limitations](#limitations)
* [Determination of URL](#determination-of-url)
* [Blocking policy](#blocking-policy)
* [URL string or blocking policy]
  * [Terminology](#terminology)
  * [Backend configuration](#backend-configuration)
  * [TXT record]
  * [Examples](#examples)
* [URL from IANA RDAP database]

## Introduction

The TLD URL feature provides a way to determine a URL to the TLD closest to a given
domain name. Zonemaster GUI makes use of this feature on its test result pages.

How the GUI gets the URL is defined in [Backend RPC API].

How Backend can override public values is defined in [Backend configuration].

This document specifies how the URL or its absence is determined.
The URL is determined based on one or more of:

* The given domain name,
* Backend configuration,
* DNS data published by the relevant TLD, and
* RDAP data published by IANA.

## Given Domain Name

In this document it is referred to the `given domain name`. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified

> here can be executed on any domain name whether or not a Zonemaster test is

Reviewer 2: ```suggestion
```

In this document the term `given domain name` is used. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be used on any domain name whether or not a Zonemaster test is
executed on the domain name.

## Preconditions

If the [given domain name] has one or more IDN labels submitted in U-label
format, then those must be converted to [A-label][RFC 5890#2.3.2.1] format before
the steps below can be run. This also applies to the TLD label.

The given domain name must be submitted in lower case, i.e. `A-Z` must be
downcased to `a-z` before submission.

The URL used is based on the [given domain name]. If the given domain name
matches one of the following conditions then no URL will be provided, and that
can not be overridden by Backend configuration.

* The given domain name is illegal, i.e. it contains illegal character(s) that

>   [prevents it from being tested by Zonemaster][Requirements]

Reviewer 2: The precondition is that the given domain name conforms to the requirements and normalization rules, right? So, we should probably just state that and link to the document. I.e., not list a selection of rules from that document and not involve logic about invalid domain names in the rules for determining the URL.

Author: I have shortened it to one sentence and refer to the document.

* The given domain name is the root zone (`.`).

> * The given domain name is a TLD, e.g. `se` or `fr`.

Reviewer 2: This part (except the item about illegal domain names) deals with the determination of the URL, so it should be moved to there.

Author: I think it fits well where it is.

## Limitations

The only types of [URLs][URL] that the mechanisms specified in this document can
result in are `http` and `https`, i.e. URLs where the scheme is `http` or
`https`. There are also restrictions on allowed characters in that URL in section
[URL string or blocking policy]. In the same section the term `URL string` is
used, and it is defined at the start of that section.

The URL used is based on the [given domain name][#given-domain-name]. If the
given domain name matches one of the following conditions then no URL will be
provided, and that can not be overridden by Backend configuration.

In all other cases a URL will be provided, if available and permitted by policy

> configuration and the restrictions in [Preconditions] are met.

Reviewer 2: The first of these paragraphs is a duplication of a paragraph in the Preconditions section that I commented that it ought to be moved to the URL determination section. The second paragraph here also deals with the URL determination, so that ought to move as well.

Author: Updated.

## Introduction

The TLD URL feature provides a way to determine a URL to the TLD closest to a given
domain name. Zonemaster GUI makes use of this feature on its test result pages.

How the GUI gets the URL is defined in [Backend RPC API].

How Backend can override public values is defined in [Backend configuration].

This document specifies how the URL or its absence is determined.
The URL is determined based on one or more of:

* The given domain name,
* Backend configuration,
* DNS data published by the relevant TLD, and
* RDAP data published by IANA.

## Given Domain Name

In this document it is referred to the `given domain name`. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be executed on any domain name whether or not a Zonemaster test is
In this document the term `given domain name` is used. The given domain name
is usually the domain name tested by Zonemaster, but the mechanisms specified
here can be used on any domain name whether or not a Zonemaster test is
executed on the domain name.

## Preconditions

> If the [given domain name][#given-domain-name] has one or more IDN labels

Reviewer 2: These links aren't recognized as links by Github.

Author: No, good catch, fixed.

result in are `http` and `https`, i.e. URLs where the scheme is `http` or
`https`. There are also restrictions on allowed characters in that URL in section
[URL string or blocking policy]. In the same section the term `URL string` is
used, and it is defined at the start of that section.

The URL used is based on the [given domain name][#given-domain-name]. If the
given domain name matches one of the following conditions then no URL will be
provided, and that can not be overridden by Backend configuration.

* The given domain name is illegal, i.e. it contains illegal character(s) that
  [prevents it from being tested by Zonemaster][Requirements]
* The given domain name is the root zone (`.`).
* The given domain name is a TLD, e.g. `se` or `fr`.

In all other cases, a URL will be provided, if available and permitted by policy
configuration.

## Determination of URL

The URL is determined based on public information, configuration and the TLD of
the given domain name. The determination order is as follows:

1. [Backend configuration] may set a specific URL for the given TLD.
2. A specific DNS record may be published by the given TLD, as specified below in

>    section [TXT record], with the URL to be used.

Reviewer 2: Maybe add `, or absence thereof, ` or something? I'm thinking that would allow the reader to get a clearer understanding of what the sub-procedure returns without having to dig through its specification.

Author: I am not sure what you mean. Please suggest change.

3. The URL for registration services found in the IANA RDAP database as specified
   below in section [URL from IANA RDAP database].

> 5. The fallback is to return the absense of a URL.

Reviewer 2: Numbering is off.

Author: Fixed.
